compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/BindsMethodValidator.kt

```diff
@@ -74,50 +95,49 @@ internal class BindsMethodValidator : PrivateCodeGenerator() {
       message = "@Binds methods must return a value (not void)",
       functionReference = function,
     )
+    val returnType = function.returnType()
 
-    if (!function.parameterMatchesReturnType() && !function.receiverMatchesReturnType()) {
-      val returnType = function.returnType().asClassReference().shortName
-      val paramSuperTypes = (function.parameterSuperTypes() ?: function.receiverSuperTypes())!!
-        .map { it.shortName }
-        .toList()
+    val parameterSuperTypes = function.singleParameterTypeOrNull().superTypes()
+    val receiverSuperTypes = function.receiverType().superTypes()
 
-      val superTypesMessage = if (paramSuperTypes.size == 1) {
-        "has no supertypes."
-      } else {
-        "only has the following supertypes: ${paramSuperTypes.drop(1)}"
-      }
-      throw AnvilCompilationExceptionFunctionReference(
-        message = "@Binds methods' parameter type must be assignable to the return type. " +
-          "Expected binding of type $returnType but impl parameter of type " +
-          "${paramSuperTypes.first()} $superTypesMessage",
-        functionReference = function,
-      )
+    when {
+      parameterSuperTypes.any { it.isAssignableTo(returnType) } -> return
+      receiverSuperTypes.any { it.isAssignableTo(returnType) } -> return
     }
-  }
 
-  private fun MemberFunctionReference.Psi.parameterMatchesReturnType(): Boolean {
-    return parameterSuperTypes()
-      ?.contains(returnType().asClassReference())
-      ?: false
```

Comment:
I would propose keeping the existing structure in the validator and modifying this method to utilize `TypeName` instead of `ClassReference` since it contains more of the information that we need in this case. You should be able to use
```
      ?.map { it.asTypeName() }
      ?.contains(returnType().asClassReference().asTypeName())
```
check for the full type info. Then `parameterSuperTypes()` would also be updated accordingly to provide `TypeReference`s.